editor/plugins/gradient_texture_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,283 @@
+		} break;
+	}
+}
+
+GradientTexture2DEditorRect::GradientTexture2DEditorRect() {
+	undo_redo = EditorNode::get_singleton()->get_undo_redo();
+
+	checkerboard = memnew(TextureRect);
+	checkerboard->set_stretch_mode(TextureRect::STRETCH_TILE);
+	checkerboard->set_draw_behind_parent(true);
+	add_child(checkerboard);
+}
+
+///////////////////////
+
+void GradientTexture2DEditor::_reverse_button_pressed() {
+	undo_redo->create_action(TTR("Swap GradientTexture2D Fill Points"));
+	undo_redo->add_do_property(texture.ptr(), "fill_from", texture->get_fill_to());
+	undo_redo->add_do_property(texture.ptr(), "fill_to", texture->get_fill_from());
+	undo_redo->add_undo_property(texture.ptr(), "fill_from", texture->get_fill_from());
+	undo_redo->add_undo_property(texture.ptr(), "fill_to", texture->get_fill_to());
+	undo_redo->commit_action();
+}
+
+void GradientTexture2DEditor::_set_snap_enabled(bool p_enabled) {
+	texture_editor_rect->set_snap_enabled(p_enabled);
+
+	snap_size_edit->set_visible(p_enabled);
+}
+
+void GradientTexture2DEditor::_set_snap_size(float p_snap_size) {
+	texture_editor_rect->set_snap_size(MAX(p_snap_size, 0.01));
+}
+
+void GradientTexture2DEditor::set_texture(Ref<GradientTexture2D> &p_texture) {
+	texture = p_texture;
+	texture_editor_rect->set_texture(p_texture);
+}
+
+void GradientTexture2DEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			reverse_button->set_icon(get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons")));
+			snap_button->set_icon(get_theme_icon(SNAME("SnapGrid"), SNAME("EditorIcons")));
+		} break;
+	}
+}
+
+GradientTexture2DEditor::GradientTexture2DEditor() {
+	undo_redo = EditorNode::get_singleton()->get_undo_redo();
+
+	HFlowContainer *toolbar = memnew(HFlowContainer);
+	add_child(toolbar);
+
+	reverse_button = memnew(Button);
+	reverse_button->set_tooltip(TTR("Swap Gradient Fill Points"));
+	toolbar->add_child(reverse_button);
+	reverse_button->connect("pressed", callable_mp(this, &GradientTexture2DEditor::_reverse_button_pressed));
+
+	toolbar->add_child(memnew(VSeparator));
+
+	snap_button = memnew(Button);
+	snap_button->set_tooltip(TTR("Toggle Grid Snap"));
+	snap_button->set_toggle_mode(true);
+	toolbar->add_child(snap_button);
+	snap_button->connect("toggled", callable_mp(this, &GradientTexture2DEditor::_set_snap_enabled));
+
+	snap_size_edit = memnew(EditorSpinSlider);
+	snap_size_edit->set_min(0.01);
+	snap_size_edit->set_max(0.5);
+	snap_size_edit->set_step(0.01);
+	snap_size_edit->set_value(0.1);
+	snap_size_edit->set_custom_minimum_size(Size2(65 * EDSCALE, 0));
+	toolbar->add_child(snap_size_edit);
+	snap_size_edit->connect("value_changed", callable_mp(this, &GradientTexture2DEditor::_set_snap_size));
+
+	texture_editor_rect = memnew(GradientTexture2DEditorRect);
+	add_child(texture_editor_rect);
+
+	set_mouse_filter(MOUSE_FILTER_STOP);
+	_set_snap_enabled(false);
```

Comment:
```suggestion
	_set_snap_enabled(snap_button->is_pressed());
```
It doesn't matter that much, but it's more consistent with snap size.